Make SAI sstable components immutable

pcmanus opened the pull request:
When a SAI index is rebuild, create new index files rather than overwriting existing files, making SAI index components
files essentially truly immutable (once initially written).

To do this, SAI index components have a new parameter, their "generation": for a given version, the component with the
highest generation is the most recently written one. That generation is added to the filename, but with the special case
that the initial generation (for a version), generation 0, is special cased and not added to the filename, and this for
backward compatibility. This means in particular that filenames are completely unchanged by this patch up until one
triggers an index rebuild that does not also correspond to a version change.

One implication is that on a rebuild, old SAI index components are not deleted anymore (previously, they were deleted at
the end of the rebuild, and the rebuild was re-creating new versions of the same files, essentially overwriting the file
but in 2 steps) and will remain on disk. The intent here is that this pave the way for allowing index "live" rebuild,
that is rebuilding an index without making the index un-queriable during the rebuild. However, this patch does *not*
implement such "live" rebuild, leaving it as a follow up. As of this patch, the old index component files are also never
deleted, though over time, compaction will get rid of them by virtue of replacing the sstables that have those
components.

Technically, the generation added is not incremented "per component", but rather per "group" of components, where
"group" means (for a particular sstable) either 1) all the per-sstable components or 2) all the components of a
particular index. As an example, suppose the components for an sstable are (where 1 is the generation, and including
only a subset of real components to keep the example shorter):
```
<sstable X>-SAI+CA+1+GroupMeta.db
<sstable X>-SAI+CA+1+TokenValues.db
<sstable X>-SAI+CA+1+PrimaryKeyTrie.db
<sstable X>-SAI+CA+1+GroupComplete.db
<sstable X>-SAI+CA+myIndex+Meta.db
<sstable X>-SAI+CA+myIndex+TermsData.db
```
In this example, the per-sstable group is complete, but not the group for `myIndex` (missing at least the
`-ColumnComplete` component). Also, the per-sstable group is at generation 1, while the `myIndex` group is at generation
0.. If a rebuild is triggered, the created components will be (assuming `CA` is the current version and again using a
subset of the actual components for simplicity):
```
<sstable X>-SAI+CA+2+GroupMeta.db
<sstable X>-SAI+CA+2+TokenValues.db
<sstable X>-SAI+CA+2+PrimaryKeyTrie.db
<sstable X>-SAI+CA+2+GroupComplete.db
<sstable X>-SAI+CA+1+myIndex+Meta.db
<sstable X>-SAI+CA+1+myIndex+TermsData.db
<sstable X>-SAI+CA+1+PostingLists.db
<sstable X>-SAI+CA+1+ColumnComplete.db
```
In other words, the generation for the per-sstable group and the `myIndex` one are independent, but the generation are
linked within a group (the `<sstable X>-SAI+CA+1+PostingLists.db` file is at generation 1 like the rest of that group
and not 0, despite there being no generation 0 file before the rebuild).

The reason generation are linked "by group" is that 1) we need to at least link the generation to the "completion
marker" components to the rest of the component of the group this is a marker for, and linking the whole group as above
feels like the cleaner solution and 2) linking the generation of *all* the SAI components of a given sstable would
prevent us from being able to rebuild a single index without rebuilding the per-sstable components/other indexes.

pcmanus commented:
Some technical details on the changes here for reviewers sake.

The core of the changes of this patch is `IndexDescriptor`, which handles the SAI components for an sstable, but that file is change fairly drastically. The rest of the changes is by and large propagating the changes to that file to all the places in the code where it is used. And that's that's probably the right place to start looking at for a review.

This patch changes the fact that there can be only a single file for a given sstable, version, component type and index "context", by adding the additional generation. This has the implication that we need to distinguish a bit more about the components for when we write a new index, from the current "active" components of an sstable. In the former case in particular (writing new index), the code used to be able to not care about what components previously existing, but with this patch, we need to know what exists and potentially bump the generation accordingly.

Additionally, as explained in the description above, the generation is linked between components of the same "group", but the existing code didn't had any abstraction for that notion of "group" of components (the notion always existed in a sense, but only abstractly), which made it really hard to make the intended changes. So the bulk of the changes in this patch is really to introduce this notion of "group" of components, where each sstable has a per-sstable "group", for all its per-sstable components, and per-index groups that, for each index, groups the per-index components of that index.

This notion of group of components in the patch is the `ComponentGroup` class (it is an interface and its only concrete implementation is a private subclass in `IndexDescriptor` in practice). It is mostly used through it's 2 sub-interfaces, `ComponentGroup.Writer` and `ComponentGroup.Reader`, whose goal is to make it more explicit in the code when we're writing new components, versus when we just read existing ones. With that, `IndexDescriptor` basically tracks the `ComponentGroup` active for an sstable, and most of the methods of `IndexDescriptor` are moved to `ComponentGroup` (which in some cases saves some duplication of code between per-sstable and per-index stuffs).

Another slight change to `IndexDescriptor` is that it always tries to "discover" the components on disk: previously, there was an `IndexDescriptor.createNew(...)` method that was used when building index and wasn't checking anything existing on disk because even if something did (because it's a rebuild), we were overwriting things anyway. With this patch however, we need to know what exists, if anything, to bump the generation if appropriate.

The "discovering" of the existing components of a sstable also has to change. Before this patch, we never "scanned" the disk for this: there is only a finite and fairly small number of versions and component kind in existence, so we just tried all combination and looked if the corresponding file existing. With generations, we can't really do that anymore, because in theory there is no saying what the most recent generation may be (in practice, it's never going to be big, there is only so may rebuild a sstable may be part of in its lifetime, but it's still hard to cap it completely). So this patch change the discovery to actually scan the sstable directory for all the files belonging to the sstable it is a descriptor of, parse the file names to see which ones are SAI components and what they are, and go from there. This discovery is in `IndexDescriptor.populateComponents`, but it rely on being able to parse a SAI component file name and that's added in `Version.tryParseFilename` (because the exact format of file names actually depends on the version).

As a minor aside, this probably make the creation of `IndexDescriptor` instances a tad more costly that before, and the existing code had a tendency of calling `IndexDescriptor.create` even when a reachable instance already existed so the patch change a few of those.

Lastly, I'll point out that this patch adds a class to represent a specific component, that is one that directly map to a single file on disk, and the patch calls this `IndexComponentInfo`. I'm only mildly a fan of that name, and in a perfect word I'd have call this just `IndexComponent`, but that name is taken for "the enum of all the kind of index components there is". I'd personally be in favor of renaming that enum to `IndexComponentKind`, and then use `IndexComponent` for the `IndexComponentInfo` of this patch, but I didn't do it in this patch because I wasn't sure if this would be "controversial". But if reviewers also thinks this is worthy renaming, happy to add it.

jasonstack commented on src/java/org/apache/cassandra/index/sai/SSTableIndex.java:
```diff
@@ -245,12 +246,12 @@ public void release()
-            }
+            var group = sstableContext.indexDescriptor.perIndexGroup(indexContext);
+            if (obsolete.get() && !group.version().useImmutableComponentFiles())
+                group.asWriter().forceDeleteAllComponents();
```
If index is dropped, writer should still be able to discard index files locally..

jasonstack replied:
Looking at it again,  are they removed by `SAI#getInvalidateTask` ?

pcmanus replied:
> are they removed by `SAI#getInvalidateTask` ?

Yeah.

To be honest, I don't think the whole `SSTableIndex::markObsolete` is still useful after this patch, but I was a bit worried of breaking something unintentionally.

More precisely, my understanding is that `SSTableIndex::markObsolete` is called in 2 cases:
1. during index build (rebuild or not), by `SAI.SAIBuildingSupport::getIndexBuildTask` (or `getParallelIndexBuildTasks`).
2. during index drop, by `SAI#getInvalidateTask`.
and when we have immutable components, we want to delete in the 2nd case but not the first one, so I added deletion directly in `SAI#getInvalidateTask`.

Now, when we do not use immutable components, we still have the added deletion in `SAI#getInvalidateTask`, so that means the deletion in `SSTableIndex::markObsolete` is only maybe useful for index build. But when we build indexes for a sstable, we already delete the old files in that case (in `SAIBuilder`), so I wonder if it's truly useful. That said, I'm only so familiar with all this code and as said above, I was worried about breaking something.

But if my reasoning above reasonates, I can remove `SSTable::markObsolete` and replace it's calls by just `SSTable::release`. That might make things overall easier to follow. WDYT?

pcmanus replied:
I actually ended up keeping `SSTableIndex::markObsolete`, though I renamed it to `markIndexDropped` to hopefully make it more explicit as to when it is called, and left the deletion of the component there, but removing from `getInvalidateTask`. That's because as you pointed out in one of your other comment, we should delete stuff only after we're sure they are released, which is what the "obsolete" part was doing.

But I also replaced the call to `markObsolete` by just `release` in `IndexViewManager::drop` (which is confusingly only called before a rebuild, so I renamed it): that's case 1 above and the builder will already get rid of components as appropriate.

jasonstack commented:
> Lastly, I'll point out that this patch adds a class to represent a specific component, that is one that directly map to a single file on disk, and the patch calls this IndexComponentInfo. I'm only mildly a fan of that name, and in a perfect word I'd have call this just IndexComponent, but that name is taken for "the enum of all the kind of index components there is". I'd personally be in favor of renaming that enum to IndexComponentKind, and then use IndexComponent for the IndexComponentInfo of this patch, but I didn't do it in this patch because I wasn't sure if this would be "controversial". But if reviewers also thinks this is worthy renaming, happy to add it.

+1 to renming. How about `IndexComponentType` or `ComponentType`?

jasonstack commented on src/java/org/apache/cassandra/index/sai/disk/v1/V1OnDiskFormat.java:
```diff
@@ -178,57 +180,26 @@ public PerIndexWriter newPerIndexWriter(StorageAttachedIndex index,
-
-            try (IndexInput input = indexDescriptor.openPerSSTableInput(indexComponent))
-            {
-                Version earliest = getExpectedEarliestVersion(indexComponent);
```
The original code seems confusing. `perSSTableComponents()` won't return vector components. There is no point using `getExpectedEarliestVersion` here...

pcmanus replied:
That was true in the original (though I don't think it was hurting, and maybe the intent was that `getExpectedEarlierVersion` could later handle more cases and so calling in all case avoided future surprises).

In the update code though, the `validateIndexComponent` method is called for both per-sstable component and per-group components, so we need the call in general (we could, of course, check `component.group().isPerIndexGroup()` if we really wanted to only call `getExpectedEarlierVersion` on per-index groups, but as mentioned above, I don't think it really hurts to call it regardless).

jasonstack commented on src/java/org/apache/cassandra/index/sai/disk/v1/V1SearchableIndex.java:
```diff
@@ -78,11 +79,12 @@ public V1SearchableIndex(SSTableContext sstableContext, IndexContext indexContex
         try
         {
-            this.indexFiles = new PerIndexFiles(sstableContext.indexDescriptor, indexContext);
+            ComponentGroup.Reader componentsGroup = sstableContext.indexDescriptor.perIndexGroup(indexContext);
```
```suggestion
            ComponentGroup.Reader perIndexGroup = sstableContext.indexDescriptor.perIndexGroup(indexContext);
```

jasonstack reviewed (commented):
Half way through..

pcmanus commented on src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java:
```diff
@@ -89,162 +99,179 @@ public class IndexDescriptor
+        String indexName = context == null ? null : context.getIndexName();
+        Map<Version, IntObjectMap<Set<IndexComponent>>> candidates = Maps.newHashMap();
+
+        PathUtils.forEach(descriptor.descriptor.directory.toPath(), path -> {
```
> How about checking the completion file (instead of listing the entire table directory) starting from generation 0 until no files are found for the generation?

I though about it, but there is 2 reasons I'm not sure this really work:
1. this assumes we do keep the files for every generation "forever" (forever in the sense of "as long as the sstable exists" which admittedly is not forever forever). And while this patch does not implement any deletion of old generation, I'd worry about not being able to at least write some async "cleanup" task for it later if we need to. I guess that cleanup task could delete all the generation file but the completion markers, but this still feel fragile, And cndb aside, I think this change is useful OSS-wise too, but relying on all the previous generation markers to be present for things to work feels particularly gratuitous fragility there: we can of course ignore that if we need to, but it's probably simpler if the implementation can work in both cases well enough.
2. the current patch actually delete completion markers when it detect a corruption to ensure that group of components does not get used, but still does not overwrite the generation in a rebuild. The idea is that even if we detect corruption in one component, there may still be other (non-corrupted) component used successfully and overridding files that are in used could still give fairly unpredictable results (which ultimately could well be worst than the corruption).

Overall, the idea here is to avoid overriding files that may still be in use concurrently from rebuilds, but those files still become unused eventually once the rebuild propagate, so I'm wary of the fragility of relying of those file sticking around no matter what.

> CNDB had some performance issues around listing files

Why is this? Isn't it a problem only if we do check remote storage?

Here, I think we'd be happy for the listing to only return what is either 1) local and 2) in the `RemoteFileCacheAttributes`, and it feel those should be fast (?). If so, we can maybe pass the sstable directory through some `StorageProvider` method that, for cndb, would decorate it with some open options that would indicate that we're happy to only use local information in the listing. Does that make sense, or am I missing the problem with listing?

pcmanus replied:
Thinking about this some more, I wonder if the simplest option isn't to add a new component (per "set of components") that stores the 'last written generation'. That component would obviously be special cases to be written without generation, so we could always read it without scanning. And if that file doesn't exists, we'd assume backward compat and generation 0 (in fact, we could dispense from writing the file at all if it's gen 0).

jasonstack replied:
> Why is this? Isn't it a problem only if we do check remote storage?
> listing to only return what is either 1) local and 2) in the RemoteFileCacheAttributes, and it feel those should be fast (?).

That's what `RemoteFileSystemProvider#newDirectoryStream` does now, both 1 and 2. I think `RemoteFileAttributesCache` is not optimized.

We can go with the current implementation and optimize later if it becomes an issue. `IndexDescriptor#create` is mainly used by writer when opening sstables and compactor. Taking a couple of milliseconds on a table with a thousand sstables should be fine.

jasonstack replied:
> Thinking about this some more, I wonder if the simplest option isn't to add a new component (per "set of components") that stores the 'last written generation'. That component would obviously be special cases to be written without generation, so we could always read it without scanning. And if that file doesn't exists, we'd assume backward compat and generation 0 (in fact, we could dispense from writing the file at all if it's gen 0).

that's TOC?

eolivelli commented:
How can we roll this out to a live cluster ?

I imagine that the procedure is:
- update the code on all the writers to a version that contains this change
- wait for the system to settle
- update the compactors with this code and enable the feature

In case of problems the first version of the files is always there (because we don't delete the first version when rebuilding the indexes) so it is safe to rollback the writers to a code that doesn't support this feature (but they will see the old version of the files, before the rebuild).

Is this correct ?

pcmanus commented:
> Is this correct ?

Yes.

pcmanus commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java:
```diff
@@ -338,42 +363,24 @@ void dropIndexSSTables(Collection<SSTableReader> ss, StorageAttachedIndex index)
         {
-            results.right.forEach(sstable -> {
-                IndexDescriptor indexDescriptor = IndexDescriptor.createFrom(sstable);
-                indexDescriptor.deletePerSSTableIndexComponents();
```
It's been moved. It's now directly in `IndexDescriptor.ComponentGroupImpl#validateComponents`: if we detect corruption there, we delete the corrupted files as appropriate.

Which I did for a couple reasons:
1. it felt cleaner to me to have the handling of corruption as close from the detection of corruption as possible. Makes it easier to have slightly more helpful messages for ones (with details on what is corrupted).
2. it happens to avoid code duplication between "per-sstable" and "per-index" stuffs. It would also save duplication if we start calling validation elsewhere.
3. I handle it slightly differently with new immutable components than in "legacy" mode. In legacy, we delete the whole "group" as soon as any component is corrupted, but with immutable, I delete only the corrupted stuff and the completion marker (the idea being, don't delete anything that may be actively read unless it's provably broken). Having that difference better encapsulated felt cleaner too.

pcmanus commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java:
```diff
@@ -250,31 +251,61 @@ public boolean handles(IndexTransaction.Type type)
+        // 0 for new sstables, but may not be for rebuilds. To handle that, we rely on the same logic used when we do
+        // write the new components (`IndexDescriptor.newPerSSTableGroupWriter`/`IndexDescriptor.newPerIndexGroupWriter`),
+        // even if we don't write anything at that point.
+        // Do note that we create an `IndexDescriptor` from scratch, and do not look in the context manager for an
```
Oops, classic me: I started the comment, though of something else, and forgot to finish it...

Pushed the end of the comment, but yes, the idea here is that this is call any time we write index components, and so half the time it will be for sstables that don't exist yet (flush or compaction). And I don't want to populate the descriptor in the `SSTableContextManager` just yet because if something fails, it wouldn't be cleared correctly.

Admittedly, half of the time the sstable will exists (rebuild, or just building a new index on an existing sstable) so we could try go get the descriptor from the context only if it exists and call `IndexDescriptor.create` directly otherwise, but this didn't felt worth the added code.

pcmanus commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java:
```diff
@@ -520,9 +517,12 @@ public Callable<?> getInvalidateTask()
+            {
+                var componentGroup = sstableIndex.getSSTableContext().indexDescriptor.perIndexGroup(getIndexContext());
+                sstableIndex.getSSTable().unregisterComponents(componentGroup.allAsCustomComponents(), baseCfs.getTracker());
+                componentGroup.asWriter().forceDeleteAllComponents();
```
Good point. As commented above, I reworked this a bit so the deletion happens after the release.

pcmanus commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java:
```diff
@@ -130,9 +129,11 @@ public void removeIndex(Index index)
+            contextManager.releaseAll(context -> {
+                var componentGroup = context.indexDescriptor.perSSTableGroup();
+                context.sstable.unregisterComponents(componentGroup.allAsCustomComponents(), baseCfs.getTracker());
+                componentGroup.asWriter().forceDeleteAllComponents();
```
Agreed here too, and as said above, reworked to delete after release.

pcmanus commented on src/java/org/apache/cassandra/index/sai/disk/format/ComponentGroup.java:
```diff
@@ -0,0 +1,264 @@
+ * build is always one more than the highest generation of any component found on disk (for the group in question, and
+ * the version we writting, usually {@link Version#LATEST}).
+ */
+public interface ComponentGroup
```
Yeah, I guess the use of *group" is maybe a bit risky given `Index.Group`.

For variables, I think I'll just move to using `components`, or maybe `perSSTableComponents`/`perIndexComponents` when it feels like it matters.

For the class name, I guess `IndexComponents` make some sense, especially if I rename `IndexComponentInfo` to just `IndexComponent`. I do like that "group" carries the notion that it's not a completely random set of components, but not really a big deal.

I'll work out those renaming.

pcmanus replied:
Yeah, I guess the use of *group" is maybe a bit risky given `Index.Group`.

For variables, I think I'll just move to using `components`, or maybe `perSSTableComponents`/`perIndexComponents` when it feels like it matters.

For the class name, I guess `IndexComponents` make some sense, especially if I rename `IndexComponentInfo` to just `IndexComponent`. I do like that "group" carries the notion that it's not a completely random set of components, but not really a big deal.

I'll work out those renaming.

pcmanus replied:
Yeah, I guess the use of *group" is maybe a bit risky given `Index.Group`.

For variables, I think I'll just move to using `components`, or maybe `perSSTableComponents`/`perIndexComponents` when it feels like it matters.

For the class name, I guess `IndexComponents` make some sense, especially if I rename `IndexComponentInfo` to just `IndexComponent`. I do like that "group" carries the notion that it's not a completely random set of components, but not really a big deal.

I'll work out those renaming.

pcmanus replied:
I've pushed a bunch of renamings along the lines of what is above. There is no variable or type called "group", unless I missed something, in what is added, so hopefully that make it less confusing. I still reference to "group" in the documentation of `IndexComponents` and a couple internal places because that's still the cleanest way to explain it imo and I think it's clear from context anyway. Feel free to have a quick look and let me know if you have better ideas for some of the changes.

pcmanus commented on src/java/org/apache/cassandra/index/sai/disk/v1/MemtableIndexWriter.java:
```diff
@@ -56,28 +55,25 @@ public class MemtableIndexWriter implements PerIndexWriter
 
-    private final IndexDescriptor indexDescriptor;
-    private final IndexContext indexContext;
+    private final ComponentGroup.Writer groupWriter;
```
Renamed those. Fwiw, I renamed `Writer` to `ForWrite` (and same for reader) because those weren't really the "writers" per-se, but rather a sort of view of the component used to open writers to it. Anyway, I'm probably thinking too hard about all this :)

pcmanus commented on src/java/org/apache/cassandra/index/sai/disk/v1/bitpack/NumericValuesWriter.java:
```diff
@@ -33,52 +31,39 @@ public class NumericValuesWriter implements Closeable
         this.metadataWriter = metadataWriter;
         this.blockSize = blockSize;
+
+        SAICodecUtils.writeHeader(output);
```
Make sense, I switched back to the previous order.

jasonstack commented on src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java:
```diff
@@ -283,328 +340,433 @@ public boolean isPerSSTableBuildComplete()
+                forceDeleteAllComponents();
+
+            groups.remove(context);
+            populateComponents(IndexDescriptor.this, context);
```
Completion marker file is removed above.  `populateComponents` will still find completion marker in TOC and populate `IndexComponentsImpl` with latest generation in the following code block:

```
                IndexComponentsImpl components = descriptor.new IndexComponentsImpl(context, version, generation);
                foundTypes.forEach(components::addOrGet);
                components.isComplete = true;
                descriptor.groups.put(context, components);
```

I was expecting it to create `generation=-1`. Is it expected?

pcmanus replied:
You're right, we need to unregister the components if they are invalid. Added it (forced to pass a few new arguments but not too bad).

jasonstack commented on src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java:
```diff
@@ -89,162 +99,217 @@ public class IndexDescriptor
+        while (true)
+        {
+            IndexComponentsImpl candidate = new IndexComponentsImpl(context, version, generationToTest);
+            if (candidate.expectedComponentsForVersion().stream().noneMatch(candidate::componentExistsOnDisk))
```
would be nice to have some logs here about generation conflict..

pcmanus replied:
Added some.

jasonstack commented on test/unit/org/apache/cassandra/index/sai/cql/RebuildWithImmutableComponentsTest.java:
```diff
@@ -0,0 +1,124 @@
+        execute("INSERT INTO %s (id, val) VALUES ('3', 'otherValue')");
+
+        flush();
+
```
would be nice to check TOC contains active SAI components after flush and rebuild

> assertTrue(SSTable.readTOC(sstable.descriptor).containsAll(group.activeComponents(sstable)));

pcmanus replied:
Added.

jasonstack commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndexGroup.java:
```diff
@@ -250,31 +261,66 @@ public boolean handles(IndexTransaction.Type type)
+        // existing stables, and create it otherwise, but it doesn't feel worth bothering here.
+        IndexDescriptor indexDescriptor = IndexDescriptor.create(descriptor, metadata);
+        Set<Component> components = indexDescriptor
+                                    .newPerSSTableComponentsForWrite()
```
Would be nice to have a test in CNDB to verify index build after index drop: 
- When an index is dropped on CNDB, writer deletes index files locally without updating archive file, so the zipped TOC remains the same. If the user creates an index again, compactor should still see the old TOC with old index component generation and should use next generation.

pcmanus replied:
Yeah. There is test for drop+recreate for the NSS case, so that will be run once we merge this to cndb, but I don't know about the non-NSS case. I'll check if there is one once I merge this and propagate the change, and I'll add a test if that's not the case to make sure it's not broken.

jasonstack reviewed (commented):
Completed 2nd pass. Left one question and some test suggestions

pcmanus commented:
As mentioned on slack, the code moved to use the TOC as a mean to find the current generation of SAI components to use, and this to avoid having to scan the whole sstable directly every time as that can be slow for CNDB. But to avoid relying on the presence of the TOC too much, the previous code that scan disk is still there as a fallback if the TOC doesn't exists: because it's only a fallback for hopefully exceptional case (we lost the TOC somehow), having it be maybe a bit slow at times shouldn't be a problem.

But it does mean that if the TOC exists, then that _is_ now relied on to know if SAI components exists or not, so hopefully existing TOC files are not broken (I don't think they are, but just pointing out that the increased reliance on them is kind of new).

I just pushed those last changed, so will wait to see if there is CI failures, but otherwise I think I've addressed all review feedback and any known issues.

jasonstack reviewed (approved):
+1

pcmanus commented:
1. I know this was approved before, but needed to make a few changes for the sake of the cndb, so re-requested review to have a look on the last changes (I've kept them in separate commit so only those probably need a look). There changes basically do the following:
1. a `reloading` method is added to re-discover things from disk. That's not used in C* proper (outside of the test I added) because any change to components is done by the done through methods of `IndexDescriptor', but in cndb, components are modified "externally" by compactor, so we need this to tell writers to re-check.
2. I've change a bit when `IndexDescriptor` does component discovery from disk. Historically, the per-sstable components were looked up when creating the `IndexDescriptor` instance and then each per-index "group" was looked up lazily, the first time any method for the index context was used. And so far I had kept that. But that lazy loading also means it's not entirely clear when we do discover components: it could well happen at any place were we check anything on per-index components, which is a lot of place, and in the cndb, with NSS, that's not great because things can change in remote storage concurrently of what any particular writer is doing, but a read of the TOC _will_ go to NSS and see whatever the last version is in remote, so that makes it really hard to reason about. Tl;dr, I've changed the code so that `IndexDescriptor` always loads everything eagerly when it is created. It means we need to pass the existing index to the ctor, but that's easy enough. As it happens, it's also a bit more efficient as it means we only read the TOC once, and not once per "group" of components (obviously doesn't matter too much, but well, still technically more efficient).
3. The case of rebuilding indexes that were at an older version to a newer one was slightly broken (we were not saving the correct components in the TOC because the wrong version was used at one point) so this fixes it and add a test for it.

eolivelli commented on test/unit/org/apache/cassandra/inject/ActionBuilder.java:
```diff
@@ -127,7 +127,7 @@ public ConditionsBuilder when(Object expression)
         public ConditionsBuilder not()
         {
-            if (!(elements.getLast() instanceof LogicOp))
+            if (!elements.isEmpty() && !(elements.getLast() instanceof LogicOp))
```
this change seem unrelated?

pcmanus replied:
It's necessary for some of the changes in the tests to not thrown a NPE. Namely, if you look at the updated injection conditions in `NodeStartupTest`, they start by a `not()` and so were hitting this (where it was trying to get the last element that just doesn't exists).

Not that this is testing code only.

eolivelli reviewed (approved):
+1

jasonstack commented on src/java/org/apache/cassandra/index/sai/disk/format/ComponentDiscovery.java:
```diff
@@ -0,0 +1,277 @@
+                    continue;
+
+                // And all start with "SAI" (the rest can depend on the version, but that part is common to all version)
+                if (!component.name.startsWith("SAI"))
```
```suggestion
                if (!component.name.startsWith(Version.SAI_DESCRIPTOR))
```

jasonstack reviewed (approved):
LGTM

pcmanus commented on test/unit/org/apache/cassandra/index/sai/SAITester.java:
```diff
@@ -334,15 +365,49 @@ protected boolean verifyChecksum(IndexContext context)
+                    // Make sure the index does use the context it should (for which we already checked the version),
+                    // but note that `SSTableIndex` uses a "sharedCopy()" of the context, so we check equality of the
+                    // pk map factory (which should be same between original and a copy).
+                    assertEquals(sstableIndex.usedPerIndexComponents(), expectedVersion);
```
Good catch. Especially since the test relying on this was passing, which was suss. Turns out the test in question was not passing the proper `IndexContext` values, so the `View` was always empty and this was never tested. Fixed that as well and double-checked this now assert things correctly.

pcmanus commented on test/unit/org/apache/cassandra/index/sai/cql/AbstractRebuildAndImmutableComponentsTest.java:
```diff
@@ -0,0 +1,82 @@
+ * with just the final validation of sstables being abstracted, but when doing that, something in CI (I suspect `ant`)
+ * tries to run the test directly on the abstract class (which fails with an `InstantiationException`, obviously, but
+ * ends up as a test failure). If there is a magic flag to make it work, I don't know it and don't care to looking for
+ * it. So the concrete tests are in the subclasses, even if it's not quite DRY.
```
I hadn't tough of that :). Seems that the renaming doesn't help, but `@Ignore` does seem to do the trick. I've reverted my work around and added `@Ignore` instead.

pcmanus commented on src/java/org/apache/cassandra/index/sai/SSTableContextManager.java:
```diff
@@ -68,42 +85,77 @@ public Pair<Set<SSTableContext>, Set<SSTableReader>> update(Collection<SSTableRe
+
+        // Now, if we create a new one, we should close the previous one if it exists.
+        if (previousContext != null)
+            previousContext.close();
```
Good call. I mentioned that in a comment, that's useful context that may be immediately obvious to future readers.
